packages/hardhat/contracts/Distributor.sol

```diff
@@ -26,49 +26,92 @@
 */
 pragma solidity >=0.8.0 <0.9.0;
 
-interface IERC20 {
-  function transfer(address to, uint256 value) external returns (bool);
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
 
-  function transferFrom(
-    address from,
-    address to,
-    uint256 value
-  ) external returns (bool);
-}
 /* 
 Implementation Resources: 
   https://github.com/banteg/disperse-research
 */
 contract Distributor {
-  event Distributed(address indexed sender, string message);
+    using SafeERC20 for IERC20;
 
-  function distributeEther(address[] memory recipients, uint256[] memory values)
-    external
-    payable
-  {
-    for (uint256 i = 0; i < recipients.length; i++) {
-      payable(recipients[i]).transfer(values[i]);
-    }
-    uint256 balance = address(this).balance;
-    if (balance > 0) {
-      payable(msg.sender).transfer(balance);
-    }
-    emit Distributed(msg.sender, "Distributed");
-  }
+    event ethDistributed(address indexed sender, string indexed name);
+    event tokenDistributed(
+        address indexed sender,
+        address indexed token,
+        string indexed name
+    );
+
+    function distributeEther(
+        address[] memory recipients,
+        uint256[] memory values,
+        string memory name
+    ) external payable {
+        require(
+            recipients.length == values.length,
+            "Invalid recipients to values ratio"
+        );
+
+        uint256 total = 0;
+        for (uint256 i = 0; i < recipients.length; i++) {
+            require(
+                recipients[i] != address(0),
+                "You can't distribute to the genesis address"
+            );
+            require(values[i] > 0, "Why waste gas for zero Ether send");
+            total += values[i];
+        }
+
+        require(msg.value >= total, "Not enough Ether to distribute");
+
+        for (uint256 i = 0; i < recipients.length; i++) {
+            (bool sent, ) = recipients[i].call{value: values[i]}("");
+            require(sent, "Failed to send Ether");
+        }
 
-  function distributeToken(
-    IERC20 token,
-    address[] memory recipients,
-    uint256[] memory values
-  ) external {
-    uint256 total = 0;
-    for (uint256 i = 0; i < recipients.length; i++) {
-      total += values[i];
+        if (msg.value > total) {
+            (bool sent, ) = msg.sender.call{value: msg.value - total}("");
+            require(sent, "Failed to refund leftover Ether");
+        }
+
+        emit ethDistributed(msg.sender, name);
     }
-    require(token.transferFrom(msg.sender, address(this), total));
-    for (uint256 i = 0; i < recipients.length; i++) {
-      require(token.transfer(recipients[i], values[i]));
+
+    function distributeToken(
+        IERC20 token,
```

Comment:
ERC20 implements the internal method ` _transfer(address sender, address recipient, uint256 amount)` with requirements about spenders balances, null addresses, and allowances. Does that make these `require`ments redundant?